emacs-the-best-python-editor/PyEval/pyeval_expression.py

```diff
@@ -0,0 +1,188 @@
+        ):
+
+            # Skip any leading whitespace characters
+            while current_char.isspace():
+                self.__current_position += 1
+                current_char = self.__expr_string[self.__current_position]
+
+            # Store whatever is next in the current_token string
+            current_token = ""
+
+            # If we are looking for an operand
+            if expecting_operand:
+                # First, we need to check for a leading '-' or '+' sign
+                if current_char == "-" or current_char == "+":
+                    current_token += current_char
+                    self.__current_position += 1
+                    current_char = self.__expr_string[self.__current_position]
+
+                # Now we loop for as long as we have numbers
+                while current_char in [
+                    "0",
+                    "1",
+                    "2",
+                    "3",
+                    "4",
+                    "5",
+                    "6",
+                    "7",
+                    "8",
+                    "9",
+                ]:
+                    current_token += current_char
+                    self.__current_position += 1
+                    current_char = self.__expr_string[self.__current_position]
+
+                # We should have a number now - add it to the output string,
+                # space delimited
+                self.__output_string += current_token + " "
+
+                # And after every operand, we need to look for an operator
+                expecting_operand = False
+
+            else:
+                # Here, we just need a single operator, so
+                # Get that operator, validate it, then
+                # Create a new operator object
+                if current_char not in ["+", "-", "*", "/", "%", "^"]:
+                    raise SyntaxError
+
+                current_operator = Operator(current_char)
+
+                # Now comes the shunting yard part
+                # - If the operator stack is empty, push the current operator
+                # - Else
+                #   - While the top of stack operator is higher precedence
+                #     - Pop it and output it.
+                #   - Push the current operator
+
+                if len(self.__operator_stack) == 0:
+                    self.__operator_stack.append(current_operator)
+
+                else:
+                    top_operator = self.__operator_stack[
+                        len(self.__operator_stack) - 1
+                    ]
+                    while (
+                        len(self.__operator_stack) > 0
+                        and top_operator.precedence
+                        > current_operator.precedence
+                    ):
+                        self.__output_string += top_operator.op_string + " "
+                        self.__operator_stack.pop()
+                        if len(self.__operator_stack) > 0:
```

Comment:
Replace this with:

```
if self.__operator_stack:
    top_operator = self.__operator_stack[-1]
```

Reply:
Lots of more Pythonic code called for! Changes in #a456bac4cf583a60c435222aab5044f8c552d4ea